### PR TITLE
Add custom network RPC method

### LINKED ISSUE #5101: API for dapp to request network change
Hello,

I would like to connect a metamask user to a custom network directly from my dapp.
The idea is to avoid the user to have to copy the new custom RPC url and paste it in a second page metamask form.

## Context :
Today when a dapp running on a custom Ethereum blockchain wants a user to sign a transaction, the user needs to :
* Copy the custom url
* Open Metamask
* Click the dropdown menu
* Click the parameters button
* Search for the « New RPC URL » form
* Paste the url
* Realize that the other form fields has no connection and are not needed
* Click the save button
* Click to the cross to go back to the main menu

And this a very long process. And metamask does not always remember the custom network.

## My proposal :
The web3 metamask provider accepts a new function to add a custom RPC from the dapp.
The metamask extension shows a window to warn the metamask user that it’s network has been changed. The warning should be enforced if the chainId is 1.
The dapp can ask the user to send a tx to the custom network using `eth_sendTransaction`.

What do you think about this feature ?

### PR DESCRIPTION
**Pull request by rekmarks**
Implement `wallet_addEthereumChain` in background, per [EIP-3085](https://eips.ethereum.org/EIPS/eip-3085).

- UI is WIP in a number of different PRs

Fixes #5101

### CODE REVIEW

File: app/scripts/controllers/permissions/enums.js
Comment (pattarapon23) [inline app/scripts/controllers/permissions/enums.js:1]: 187dca2

**Review by pattarapon23 — APPROVED** [review]

**Review by pattarapon23 — DISMISSED** [review]

**Comment by Gudahtt** [comment]
I believe this addresses #5101

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:56]: Hm, I'm thinking about the benefits of this. Why shouldn't we just ignore them?

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:93]: We should do both of these before merging this.

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:85]: Here, we should now call `isSafeChainId` as done in #10224.

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:133]: - We can move the `.decimals` and `.symbol` validation into a single `nativeCurrency !== null` condition.
- We can just do `nativeCurrency.decimals !== 18`

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:138]: Any thoughts on how long the ticker should be? IMO, it's really about what we're able to handle in the UI, but also about what's sensible for a native currency.

- Maybe 6 as max is better?
- Should we allow tickers of length 1?

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:141]: This error message needs to be updated either way.

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:161]: These seem like cheap checks that should happen earlier.

**Review by rekmarks — COMMENTED** [review]
TIL "Pull request authors can't request changes on their own pull request". 👍

**Comment by brad-decker** [comment]
@rekmarks responded to your feedback

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (brad-decker) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js]: done in latest

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (brad-decker) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js]: Done

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (brad-decker) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js]: update allows for between 2 and 6 chars

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (brad-decker) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js]: done

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (brad-decker) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:88]: done

**Comment by vincentlg** [comment]
I would like to have clarification on the status of this PR and possibly help to finish it if possible. I find this feature absolutely essential to facilitate the connection to L2.
I tested build locally (chrome / firefox) but nothing happens with these parameters (except badge "1" request awaiting confirmation on the Metamask icon)

```
method: "wallet_addEthereumChain",
params: [
  {
    "chainId": "0x64",
    "chainName": "xDAI Chain",
    "rpcUrls": ["https://dai.poa.network"],
    "iconUrl": "https://xdaichain.com/fake/example/url/xdai.svg",
    "nativeCurrency": {
      "name": "xDAI",
      "symbol": "xDAI",
      "decimals": 18
    }
  }
]
```
Is this normal considering the "draft" status or am I testing in the wrong way? 
What is the expected behavior in terms of UI?  
Thank you!

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:98]: nit: Given that `_chainId` can be `false`, it makes more sense for this check to happen after `isPrefixedFormattedHexString` and `isSafeChainId`.

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:154]: nit: the `typeof nativeCurrency !== 'object'` can be brought into the `if` block currently starting on line 156.

I also think the complete check should be: 

`typeof nativeCurrency !== 'object' || Array.isArray(nativeCurrency)`

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:60]: That makes sense, let's keep it!

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js]: You mean `customRpcExistsWith`? That does not cover the default networks.

**Comment by rekmarks** [comment]
Hi @vincentlg, thanks for reaching out! The behavior you're describing is expected on this branch, which only includes the background application logic for this feature. @brad-decker has the UI in progress [on this other branch] (https://github.com/MetaMask/metamask-extension/tree/custom-network-rpc-ui), which may or may not be up-to-date with the latest work here.

I wish I had a way for you to contribute, but all work for implementing EIP-3085 in MetaMask is already nearing completion. Please let me know if there's anything else I can do for you!

**Comment by vincentlg** [comment]
Thanks @rekmarks  for the details, any idea when this will be released? can't wait to test this feature for our users ;)

**Comment by rekmarks** [comment]
@vincentlg it's a current priority and the next new feature on our roadmap, so Very Soon(TM) 😉

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:3]:
```suggestion
```

**Comment by rekmarks** [comment]
This looks good to me. Discussing whether to stack the UI implementation on top of this PR or merge this first. Will move out of draft regardless, but leave the "DONOTMERGE" label for now.

**Comment by brad-decker** [comment]
I asked during a standup what we'd like to do, and I thought the comments from @danjm were to allow this to land but toggled behind a flag or something. If we stack the UI pr on top of this one, it would be nontraditional to how we usually handle stacked PRs. You'd want to review each separately, and then as a whole, before merging.

My preference would be to comment out the wiring up of this method such that the code is unreachable and then remove the comment in the UI PR.

**Comment by brad-decker** [comment]
@Gudahtt do you have a preference? Considering neither @rekmarks or I can review this PR now 😅

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (rekmarks) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:48]: I just remembered: It was discussed on EIP-3085 that this field should be changed to `iconUrls` (plural) and be string array. I'm going to update the EIP to encompass that.

We don't currently do anything with this property anyway as far as I can tell, but just, FYI.

**Comment by Gudahtt** [comment]
Ideally each PR would be:
1. Non-breaking
2. Easy to review
3. Useful (i.e. no dead code)

For larger interconnected changes, I guess it's a question of sacrificing 2 or 3. If we review the changes separately and then merge it all in a monster PR we kinda still need to review the thing as a whole, so that isn't so easy to review. But if we merge it in in smaller chunks that are unused until the final PR is merged, we end up with dead code in the interim at least. There's also a substantial review burden on the final PR to ensure there is no dead code leftover, and to ensure everything works together as expected. Probably less of a burden than the other strategy though, so generally this sounds better to me.

The "feature flag" idea seems entirely unrelated to this problem. A feature flag (whether it be a build-time flag or a user-facing toggle somewhere) can be used to trigger a change in behaviour, or to enable a new feature entirely. But that wouldn't solve the problem of there being dead code somewhere after this PR lands but before the UI code lands. It wouldn't ease any review burden either. It's a solution for an entirely different problem: when a new feature or a change in behaviour is in-progress, and we want to land it in `develop` before it's ready for production, so that we can further refine it and test it more easily while minimizing the impact of other unrelated work being done on `develop`.

>My preference would be to comment out the wiring up of this method such that the code is unreachable and then remove the comment in the UI PR.

That works!

**Comment by brad-decker** [comment]
Feature flags can and should be used to land parts of code, not intended for consumption, into the main branch without the mess of long-lived feature branches. In some cases, this will mean dead code is merged in to ease the burden of both reviewer and feature team, who would otherwise have to manage review and rebasing of highly divergent code branches. I agree that it's overkill for this case, but feature flags do not have to be reserved only for usable code. Once review on this is done, I'll comment out the wiring up of this method so that it can be uncommented out in the UI branch. Also, at that time I'll remove the Do not merge label and re-request review.

**Comment by Gudahtt** [comment]
>Feature flags can and should be used to land parts of code, not intended for consumption

Just to be clear, this is what I meant :sweat_smile: 

I was saying that feature flags are for landing things that you don't actually want in the application _by default_. They aren't a solution for how to merge a feature that is too large to review in one PR - that's a separate problem. I wasn't even saying that a feature flag would be overkill for this feature, just that it's not a solution to the problem being discussed here.

Whether we use a feature flag here or not, we still don't want the feature to be broken, and we still want to minimize review burden and dead code, and we'll still have to compromise on one or both of those to merge this in.

**Comment by brad-decker** [comment]
👍 -- we can continue the conversation on how/when we should use feature flags if the opportunity to use them ever arises. I think, considering the successive nature of the PRs, the impact of the temporary dead code (from commenting out the wiring up of the method) should be minimal.

**Comment by brad-decker** [comment]
@rekmarks https://github.com/MetaMask/metamask-extension/pull/9724/commits/335b9abce10eb65b0000f68674d6f739ae091189 implements the switch network approval. This is the only change since you last looked.

**Comment by brad-decker** [comment]
@danjm @Gudahtt -- I believe this is ready for review and if found to be good I'll comment out the code that makes the API active. Once we're ready to launch the feature I'll do a small PR to add that code back in.

File: ui/app/helpers/utils/util.js
Comment (danjm) [inline ui/app/helpers/utils/util.js:470]: Review note: this method has been moved to `rpc.utils.js` but it is unchanged

File: shared/modules/rpc.utils.js
Comment (danjm) [inline shared/modules/rpc.utils.js:13]: Review note: this is the same as the `jsonRpcRequest` that was removed from `ui/app/helpers/utils/util.js`, the code is unchanged

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (danjm) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:54]: I notice that `'iconUrls'` is not included in `otherKeys`, but neither is it used anywhere. Is there some reason we are allowing this in the request params even though it will go unused?

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (danjm) [inline app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:139]: Are we unable to support the updating of the block-explorer url?

**Review by danjm — CHANGES_REQUESTED** [review]
I left a couple comments and questions

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (brad-decker) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:139]: when we get here the user will have already added the chain to their MetaMask, and the block explorer URL will have been set. This requestData is set because updateRpcTarget requires it.

File: app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js
Comment (brad-decker) [reply app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js:54]: It is a part of EIP-3085, the goal here being to warn for extraneous keys that are not a part of the EIP. For example: 'blockExplorerUrls' is what we want, but someone might put 'blockExplorerUrl'. Beings as blockExplorerUrls is optional, without this check we'd accept the request and add the chain to the user's MetaMask without the blockExplorer, and the dapp would be none the wiser.

**Review by danjm — DISMISSED** [review]
this looks good to me

**Review by danjm — APPROVED** [review]

**Pull request merged**
